### PR TITLE
Integrate Snpe lib as Snpe Execution Provider

### PR DESCRIPTION
**Pull request by HectorSVC**
Integrate Snpe lib as Snpe Execution Provider

### CODE REVIEW

File: cmake/onnxruntime_providers.cmake [hunk 1]
```diff
@@ -753,6 +756,27 @@ if (onnxruntime_USE_NNAPI_BUILTIN)
...
+  source_group(TREE ${ONNXRUNTIME_ROOT}/core FILES ${onnxruntime_providers_snpe_cc_srcs})
+  add_library(onnxruntime_providers_snpe ${onnxruntime_providers_snpe_cc_srcs})
+  onnxruntime_add_include_to_target(onnxruntime_providers_snpe onnxruntime_common onnxruntime_framework onnx onnx_proto protobuf::libprotobuf-lite flatbuffers)
+  target_link_libraries(onnxruntime_providers_snpe PRIVATE SNPE.lib)
+  add_dependencies(onnxruntime_providers_snpe onnx ${onnxruntime_EXTERNAL_DEPENDENCIES})
+  set_target_properties(onnxruntime_providers_snpe PROPERTIES CXX_STANDARD_REQUIRED ON)
+  set_target_properties(onnxruntime_providers_snpe PROPERTIES FOLDER "ONNXRuntime")
+  target_include_directories(onnxruntime_providers_snpe PRIVATE ${ONNXRUNTIME_ROOT} ${SNPE_ROOT}/include/zdl)
+  set_target_properties(onnxruntime_providers_snpe PROPERTIES LINKER_LANGUAGE CXX)
+  install(DIRECTORY ${PROJECT_SOURCE_DIR}/../include/onnxruntime/core/providers/snpe  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/onnxruntime/core/providers)
```
Comment (jywu-msft) [inline cmake/onnxruntime_providers.cmake:777]: if the factory header is not needed for snpe, then this line isn't needed right?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -2684,6 +2684,28 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
         updateOutputElemType(ctx, 0, ONNX_NAMESPACE::TensorProto::BOOL);
       });
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Snpe)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc("Onnx node for SNPE.")
+      .Attr("payload", "payload of the SNPE DLC file.", AttributeProto::STRING)
```
Comment (jywu-msft) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2691]: prefer a different name for this attribute. how about DLC ?

File: onnxruntime/core/providers/snpe/SnpeLib.cc [hunk 3]
```diff
@@ -0,0 +1,212 @@
+#ifdef _WIN32
+#pragma warning(push)
+#pragma warning(disable : 4244)
+#pragma warning(disable : 4541)
+#endif
+
+#ifndef _WIN32
+#define dynamic_cast static_cast
```
Comment (jywu-msft) [inline onnxruntime/core/providers/snpe/SnpeLib.cc:8]: why is this needed?

File: onnxruntime/core/providers/snpe/snpe.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      auto output = info.node().OutputDefs().at(output_i);
+      auto output_shape = output->Shape();
+      for (int i = 0; i < output_shape->dim_size(); ++i) {
+        output_dims_.at(output_i).push_back(output_shape->dim(i).dim_value());
+      }
+    }
+
+    const auto payload = info.GetAttrOrDefault<std::string>("payload", "");
+    ORT_ENFORCE((payload.length() > 0), "dlc model payload is empty!");
+    const auto snpe_Ep = static_cast<const SNPEExecutionProvider*>(info.GetExecutionProvider());
```
Comment (jywu-msft) [inline onnxruntime/core/providers/snpe/snpe.h:31]: don't mix case for variables?
snpe_ep

File: cmake/onnxruntime_providers.cmake (hunk 1 above)
Comment (HectorSVC) [reply cmake/onnxruntime_providers.cmake:777]: user need to call OrtSessionOptionsAppendExecutionProvider_SNPE to create Snpe EP, I don't understand why it's not required?

File: onnxruntime/core/providers/snpe/SnpeLib.cc (hunk 3 above)
Comment (HectorSVC) [reply onnxruntime/core/providers/snpe/SnpeLib.cc:8]: removed

File: cmake/onnxruntime_providers.cmake (hunk 1 above)
Comment (jywu-msft) [reply cmake/onnxruntime_providers.cmake:777]: see pranav's comment above. all new EP factory functions go directly into c api header.

File: include/onnxruntime/core/providers/snpe/snpe_provider_factory.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the MIT License.
+#pragma once
+
+#include "onnxruntime_c_api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+ORT_EXPORT ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_SNPE,
```
[the remark this answers is not in the document]
Comment (HectorSVC) [reply include/onnxruntime/core/providers/snpe/snpe_provider_factory.h:11]: do you mean I should add this API to onnxruntime_c_api.h?
also I do see a high potential that the API will need to be extended to allow more runtime options.

**Comment by HectorSVC** [comment]
not yet, we can setup a small dlc model for the test

---
In reply to: [839077751](https://github.com/microsoft/onnxruntime/pull/7643#issuecomment-839077751) [](http://example.com/codeflow?ancestors=839077751,839077751)

File: include/onnxruntime/core/providers/snpe/snpe_provider_factory.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (jywu-msft) [reply include/onnxruntime/core/providers/snpe/snpe_provider_factory.h:11]: what other options do you think will be needed? if we can add them now, we should. ideally we shouldn't need to make many more changes to SNPE EP.
@pranavsharma  , we're kind of in a transition period here for EP options. I'm hesitant to recommend to add string key/value api now, before the full pattern has been established.  What should the recommendation be for adding options for the upcoming release?

[the remark this answers is not in the document]
Comment (pranavsharma) [reply include/onnxruntime/core/providers/snpe/snpe_provider_factory.h:11]: If you don't anticipate any more options, the current API is fine. It just needs to be moved to the C API header file. If we anticipate more options, we should go the key/value route we discussed.
